pysmt/oracles.py

```diff
@@ -174,6 +176,7 @@ def _theory_from_type(self, ty):
     @walkers.handles(op.RELATIONS)
     @walkers.handles(op.BOOL_OPERATORS)
     @walkers.handles(op.BV_OPERATORS)
+    @walkers.handles(op.STR_OPERATORS)
```

Comment:
I am not that sure. While in principle the return type of those operations is Integer, they do not require any theory reasoning at the integer level. For example, would you say that this example is in QF_SLIA or QF_S: ```(int_to_str (str_to_int 5))```? If there is no theory operator in the formula, there is no need for theory reasoning, since the constants can be considered uninterpreted.

Reply:
Not sure, but for the `ToReal` case we added both `LIA` and `LRA` (https://github.com/pysmt/pysmt/blob/dfd7944740ae60bf91e9e40df32fcc5873fe3b7e/pysmt/oracles.py#L236)

Also, I am not convinced that in the `(int_to_str (str_to_int "5"))` case you can treat the two operators as uninterpreted:
```lisp
(= (int_to_str (str_to_int "5")) "5")  ;; Is a tautology on QF_SLIA
```
while
```
(= (f (g "5")) "5") ;; Is sat, but not tautology on QF_S
```

Reply:
I think we need to agree on what is required for a formula to be in a certain theory. My understanding of what you are saying is that the use of the sort associated with the theory is sufficient. I think that using interpreted functions of the theory is the minimum requirement. The elements of the sort (e.g., Integer) can be used within uninterpreted functions without the need of involving the theory solver.

As you point out, ToReal follows the first approach, and I did the same for BV2Nat. However, we are not consistent in the ``walk_function``:

https://github.com/pysmt/pysmt/blob/dfd7944740ae60bf91e9e40df32fcc5873fe3b7e/pysmt/oracles.py#L219-L231

Since we should check the return type of the function and update the theory accordingly. I am open to both approaches. While I believe that having the smaller theory set is an advantage, in practice I would expect that in 90% of the cases if you use ToReal, BV2Nat, or StrToInt, you really want to use the arithmetic theory too!

Reply:
I fully agree.

I tested the following SMT problem on Z3 (online on rise4fun):
```lisp
(set-logic QF_UF)
(declare-fun f (Int) Int)
(assert true)
(check-sat)
(get-model)
```
And I get: `Z3(2, 16): ERROR: Parsing function declaration. Expecting sort list '(': unknown sort 'Int'`

Changing the logic in `QF_UFLIA` correctly gives a model with no errors.

Given this little experiment, I think that going for the "wider" approach where also the returning type of functions is considered to determine the theory is better.